docs/component-api.md

```diff
@@ -0,0 +1,112 @@
+
+# Component API definition and use
+
+We have chosen as Nunjucks as the templating language for GOV.UK Frontend components. We expose those templates as reusable chunks of code: macros. Developers import macros into their application, call them as per documentation and provide data to its parameters.
+
+To provide a level of consistency for developers we have standardised parameter names, their expected input, use and placement. There are expectations, and  if so they are documented accordingly.
+
+## Specifying content
+When providing *content* to a macro, say for a label or a button, we accept two parameter options:
+
+ - `text` accepts a plain string and is the default way of passing content
+ - `html` accepts html markup. In the template we will not escape html so it will be rendered. In a scenario where both text and html are set, html parameter will take precedence over text.
+
+Example:
+
+`{{ govukButton({"text": "Button text"}) }}`
+
+`{{ govukButton({"html": "Button <span class='bold'>text</span>"}) }}`
+
+Example of implementing logic in a component template:
+
+`{{ params.html | safe if params.html else params.text }}`
+
+Example shows that if `html` and `text` parameters are present, then `html` takes precedence over `text` and we are not escaping it.
+
+## Naming attributes
+
+We should use **camelCase** for naming attributes.
+
+If a component depends on another component, we group the attributes for the dependent component inside an object, where the name of the object is the name of the component using **camelCase** convention. In case of ambiguity we prefix the component name.
+
+Example of a component depending on another component
+```
+{{ govukLabel({
+	"text": "Label text",
+	"errorMessage": {
+		"text": "Error message"
+	}
+}) }}
+```
+
+Example of a component depending on two other components
+```
+{{ govukInput({
+	"name": "example-input",
+	"label": {
+		"text": "Label text"
+	},
+	"errorMessage": {
+		"text": "Error message"
+	}
+}) }}
+```
+
+## Mimic HTML attribute names
+When there is a need to specify html attributes, such as *checked, disabled, id, name*, etc, and they map directly, we use the same parameter name. We use boolean value to check and render the attribute.
+
+Example:
+
+`{{ govukButton({"disabled": true}) }}`
+
+`{{ govukCheckbox({"checked": true}) }}`
+
+
+## Defining additional HTML attributes
+When there is a need to add additional attributes to the component, we accept an ***"attributes"*** object with key : value pairs for each attribute.
+
```

Comment:
Like so?
These should not be used to overwrite already declared attributes, such as class. If there is a need for that, make use of "classes" parameter.

Reply:
How about:

> You cannot use this to set attributes that are already defined, such as `class` – use the "classes" argument instead.

?

We use 'arguments' in the README for the table, so we should use that term consistently (vs parameters, attributes, etc)

Reply:
sounds good. will add it and check where else we use parameters and update it to arguments